tensorflow-framework/src/test/java/org/tensorflow/framework/utils/ND.java

```diff
@@ -647,6 +731,97 @@ public static FloatNdArray sum(FloatNdArray a, Integer[] axes, boolean keepDims)
     }
   }
 
+  /**
+   * Sum all elements of an array
+   *
+   * @param a the array
+   * @return an a array with one element containing the sum.
+   */
+  public static DoubleNdArray sum(DoubleNdArray a) {
+    AtomicReference<Double> sum = new AtomicReference<>(0.);
+    a.scalars().forEach(f -> sum.set(sum.get() + f.getDouble()));
+    return NdArrays.scalarOf(sum.get());
+  }
+
+  /**
+   * Sum all elements of an array based on the specified axis
+   *
+   * @param a the array
+   * @param axis the axis to sum
+   * @return an a array the sum over the axis less the diemsnion
+   */
+  public static DoubleNdArray sum(DoubleNdArray a, int axis) {
+    return sum(a, axis, false);
+  }
+
+  /**
+   * Sum all elements of an array based on the specified axis
+   *
+   * @param a the array
+   * @param axis the axis to sum
+   * @param keepDims indicates whether the dimensions over the sum should be kept or not.
+   * @return an a array the sum over the axis
+   */
+  public static DoubleNdArray sum(DoubleNdArray a, int axis, boolean keepDims) {
+    Shape shape = a.shape();
+    int nDims = shape.numDimensions();
+    int xis = nDims - 1 - axis;
+    long totalSize = shape.size();
+    long axisSize = shape.size(xis);
+    final double[] sums = new double[(int) axisSize];
+
+    a.scalars()
+            .forEachIndexed(
+                    (idx, f) -> {
+                      sums[(int) idx[xis]] += f.getDouble();
+                    });
+
+    if (keepDims) {
+      long[] newDims = shape.asArray();
+      newDims[axis] = 1;
+      final AtomicInteger counter = new AtomicInteger();
```

Comment:
Why the `AtomicInteger` here? It is local.

Reply:
It is referenced in the lambda on lines 784-789, you can't use `int i` because it is not final, and if it were final you couldn't increment it.

Reply:
👍 Oh, right!